File: examples/chapter8/table_8-4.py
import warnings
warnings.filterwarnings("ignore")

import numpy as np
from sklearn.metrics import confusion_matrix, mean_squared_error
from tabulate import tabulate

from quadratic_with_cross_validation import quadratic_discrimination
from pls_matrices import pls_matrices
import functools as ft

np.random.seed(5)

FITR_Spectra_instant_coffee = np.genfromtxt("data/chapter8/coffee.csv", delimiter=",")

Y = (FITR_Spectra_instant_coffee[1, 1:57].T - 1).reshape(-1, 1)
X = (FITR_Spectra_instant_coffee[3:289, 1:57].T)

Dmax = 12
K = 10
h = np.linspace(0, 1, 6)

quad_1 = np.zeros((Dmax, len(h)))
quad_2 = np.zeros((Dmax, len(h)))

print("Performing crossvalidation on coffee data.")
for d in range(1, Dmax+1):
    for s in range(len(h)):

        print(f"{d}, {s+1} / {Dmax}, {len(h)}")
        alp = h[s]

        aux = quadratic_discrimination(X, Y, K, d, alp)

        li_1 = aux['predict_quad_1']
        li_2 = aux['predict_quad_2']

        quad_1[d-1, s] = mean_squared_error(Y, li_1)
        quad_2[d-1, s] = mean_squared_error(Y, li_2)

M1 = np.where(quad_1 == np.min(quad_1))
if len(M1[0]) <= 1:
    M1 = M1[0][0]+1, M1[1][0]
else:
    M1 = M1[0]
    M1 = M1[0]+1, M1[1]

M2 = np.where(quad_2 == np.min(quad_2))
if len(M2[0]) <= 1:
    M2 = M2[0][0]+1, M2[1][0]
else:
    M2 = M2[0]
    M2 = M2[0]+1, M2[1]

# Table 8.4
d = M1[0]
alp = M1[1]

print("Calculating LOOCV for AN-Q1 (may take a long time)")
aux = quadratic_discrimination(X, Y, X.shape[0], d, alp, loocv=True)['predict_quad_1']
cm=confusion_matrix(y_true=Y, y_pred=aux)
canq1 = cm.diagonal().sum() / cm.sum()

d = M2[0]
alp = M2[1]

print("Calculating LOOCV for AN-Q2 (may take a long time)")
aux = quadratic_discrimination(X, Y, X.shape[0], d, alp, loocv=True)['predict_quad_2']
cm=confusion_matrix(y_true=Y, y_pred=aux)
canq2 = cm.diagonal().sum() / cm.sum()


oil = np.genfromtxt("data/chapter8/FTIR_Spectra_olive_oils.csv", delimiter=",")

Y = (oil[1, 1:121].T - 1).reshape(-1, 1)
X = (oil[3:571, 1:121].T)

Dmax = 14
K = 10
h = np.linspace(0, 1, 6)

quad_1 = np.zeros((Dmax, len(h)))
quad_2 = np.zeros((Dmax, len(h)))

print("Performing crossvalidation on olive oil data.")
for d in range(1, Dmax+1):
    for s in range(len(h)):

        print(f"{d}, {s+1} / {Dmax}, {len(h)}")
        alp = h[s]

        aux = quadratic_discrimination(X, Y, K, d, alp)

        li_1 = aux['predict_quad_1']
        li_2 = aux['predict_quad_2']

        quad_1[d-1, s] = mean_squared_error(Y, li_1)
        quad_2[d-1, s] = mean_squared_error(Y, li_2)

M1 = np.where(quad_1 == np.min(quad_1))
if len(M1[0]) <= 1:
    M1 = M1[0][0]+1, M1[1][0]
else:
    M1 = M1[0]
    M1 = M1[0]+1, M1[1]

M2 = np.where(quad_2 == np.min(quad_2))
if len(M2[0]) <= 1:
    M2 = M2[0][0]+1, M2[1][0]
else:
    M2 = M2[0]
    M2 = M2[0]+1, M2[1]

d = M1[0]
alp = M1[1]

print("Calculating LOOCV for AN-Q1 (may take a long time)")
aux = quadratic_discrimination(X, Y, X.shape[0], d, alp, loocv=True)['predict_quad_1']
cm=confusion_matrix(y_true=Y, y_pred=aux)
oanq1 = cm.diagonal().sum() / cm.sum()

d = M2[0]
alp = M2[1]

print("Calculating LOOCV for AN-Q2 (may take a long time)")
aux = quadratic_discrimination(X, Y, X.shape[0], d, alp, loocv=True)['predict_quad_2']
cm=confusion_matrix(y_true=Y, y_pred=aux)
oanq2 = cm.diagonal().sum() / cm.sum()


# Table 8.4 (FULL)
table = {
    "Dataset": [
        "Coffee", "Olive Oil"
    ],
    "AN-Q1": [
        round(100*(canq1), 2), round(100*(oanq1), 2)
    ],
    "AN-Q2": [
        round(100*(canq2), 2), round(100*(oanq2), 2)
    ]
}

print("Table 8.4")
print(tabulate(table, headers=table.keys()))

File: examples/chapter8/table_8-4.R
### tomo los datos...
rm(list=ls())
library(caret)
  
library(readr)

 
FTIR_Spectra_instant_coffee <- read.csv("data/chapter8/coffee.csv", header=FALSE)
 
source("examples/chapter8/lib/quadratic_with_cross_validation.R")
source("examples/chapter8/lib/pls_matrices.R")
Y=as.numeric(t(FTIR_Spectra_instant_coffee[2,2:57]))
 
Y=matrix(Y,ncol=1)
xx=t(FTIR_Spectra_instant_coffee[4:289,2:57])

 
 

X=matrix(0,ncol=ncol(xx),nrow=nrow(xx))
for (i in 1:ncol(xx)){
  for (j in 1:nrow(xx)){
    X[j,i]=as.numeric(xx[j,i])
  }
}



Dmax=12 

K=10
h=seq(0,1,.2) 

quad_1=matrix(0,ncol=length(h),nrow=Dmax)
quad_2=matrix(0,ncol=length(h),nrow=Dmax)

set.seed(1)

cat("Fitting coffee data\n")
for (d in 1:Dmax){
  for (s in 1:length(h)){
    cat(d, ", ", s, " / ", Dmax,", ", length(h), '\n', sep="")
    
    alp=h[s]
    aux=quadratic.discrimination(X,Y,K,d,alp)
    li_1=aux$predict_quad_1
    li_2=aux$predict_quad_2
    
    quad_1[d,s]=mean((li_1-Y)^2)
    quad_2[d,s]=mean((li_2-Y)^2)
  }
}


M1=which(quad_1 == min(quad_1), arr.ind = TRUE)
M2=which(quad_2 == min(quad_2), arr.ind = TRUE)
#minimums
#quad 1 d=4, alp=0.8
#quad_2 d=2, alp =  1 


####Table 8.4: Olive oil and Coffee data: 
###Estimates of the correct classification rates (%) from leave one out cross validation. Results in columns 4–8 are as described in Table 7.1.

d=M1[1]
alp=M1[2]

aux=quadratic.discrimination(X,Y,nrow(X),d,alp)
cofanq1 <- confusionMatrix(as.factor(aux$predict_quad_1),as.factor(Y))$overall[1]


d=M2[1]
alp=M2[2]

aux=quadratic.discrimination(X,Y,nrow(X),d,alp)
cofanq2 = confusionMatrix(as.factor(aux$predict_quad_2),as.factor(Y))$overall[1]





oils <- read_csv("data/chapter8/FTIR_Spectra_olive_oils.csv", col_names = FALSE)

FTIR_Spectra_instant_coffee=oils


xx=t(FTIR_Spectra_instant_coffee[4:573,2:121])

X=matrix(0,ncol=ncol(xx),nrow=nrow(xx))
for (i in 1:ncol(xx)){
  for (j in 1:nrow(xx)){
    X[j,i]=as.numeric(xx[j,i])
  }
}



Y=as.factor(t(FTIR_Spectra_instant_coffee[3,2:121]))
Y=as.numeric(Y)
Y=matrix(Y,ncol=1)

Dmax=14
K=nrow(X)

h=seq(0,1,.2) 

quad_1=matrix(0,ncol=length(h),nrow=Dmax)
quad_2=matrix(0,ncol=length(h),nrow=Dmax)

cat("Fitting olive oil data\n")
for (d in 1:Dmax){
  for (s in 1:length(h)){
    cat(d, ", ", s, " / ", Dmax,", ", length(h), '\n', sep="")
    
    alp=h[s]
    
    aux=quadratic.discrimination(X,Y,K,d,alp)
    
    li_1=aux$predict_quad_1
    li_2=aux$predict_quad_2
    
    
    quad_1[d,s]=mean((li_1-Y)^2)
    quad_2[d,s]=mean((li_2-Y)^2)
    
    
  }
  
}



M1=which(quad_1 == min(quad_1), arr.ind = TRUE)
M2=which(quad_2 == min(quad_2), arr.ind = TRUE)

#quad_2 d= 10, 1 
#quad 1 d= 13 1

aux=quadratic.discrimination(X,Y,nrow(X),M1[1],h[M1[2]])

library(matrixcalc)

oanq1 = confusionMatrix(as.factor(aux$predict_quad_1),as.factor(Y))$overall[1]

aux=quadratic.discrimination(X,Y,nrow(X),M2[1],h[M2[2]])

oanq2<- confusionMatrix(as.factor(aux$predict_quad_2),as.factor(Y))$overall[1]


cat("Table 8.4\n")
cat("Dataset   | AN-Q1 | AN-Q2\n")
cat("----------|-------|------\n")
cat("Coffee    |", 100*cofanq1,"|", 100*cofanq2,"\n")
cat("Olive Oil |", 100*oanq1, "|", 100*oanq2, "\n")
